Jaeger: Search traces

zoltanbedi opened the pull request:
**What this PR does / why we need it**:

This PR implements search query for Jaeger.

### TODO

- [x] Add images to docs
- [x] Update UI once Nico is ready with his feedback
- [x] Add tests

**Which issue(s) this PR fixes**:

Fixes #24953
Fixes #27245

zoltanbedi committed:
WIP: Search jaeger traces

zoltanbedi committed:
Merge remote-tracking branch 'origin/master' into zoltan/better-jaeger-search

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=32805) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=32805) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=32805) it.</sub>

zoltanbedi committed:
Add more customizable query params

zoltanbedi committed:
Merge remote-tracking branch 'origin/master' into zoltan/better-jaeger-search

zoltanbedi committed:
Fix failing test

zoltanbedi committed:
Fix e2e test

zoltanbedi committed:
Add tags input field

zoltanbedi committed:
Merge remote-tracking branch 'origin/master' into zoltan/better-jaeger-search

zoltanbedi committed:
Minor changes

zoltanbedi committed:
Merge remote-tracking branch 'origin/master' into zoltan/better-jaeger-search

zoltanbedi committed:
Fix tests

zoltanbedi committed:
Add docs

zoltanbedi committed:
Make sure linking is working

zoltanbedi committed:
Add tests to datasource.ts

achatterjee-grafana reviewed (commented):
Added some review comments.

achatterjee-grafana commented on docs/sources/datasources/jaeger.md:
```diff
@@ -39,25 +39,24 @@ This is a configuration for the [trace to logs feature]({{< relref "../explore/t
 ## Configure the data source with provisioning
 
-You can set up the data source via configuration files with Grafana’s provisioning system. Refer to [provisioning docs page]({{< relref "../administration/provisioning/#datasources" >}}) for information on various settings and how it works.
+You can set up the data source via configuration files with Grafana's provisioning system. Refer to [provisioning docs page]({{< relref "../administration/provisioning/#datasources" >}}) for information on various settings and how it works.
```
```suggestion
You can set up the data source via configuration files with Grafana's provisioning system. Refer to [provisioning docs page]({{< relref "../administration/provisioning/#datasources" >}}) for more information on configuring various settings.
```

achatterjee-grafana reviewed (commented):
Another suggestion.

achatterjee-grafana commented on docs/sources/datasources/jaeger.md:
```diff
@@ -39,25 +39,24 @@ This is a configuration for the [trace to logs feature]({{< relref "../explore/t
+1. Service - Lists services.
+1. Operation - Populated after selecting a service. Lists operations related to the selected service. Select [All] option to query all the operation.
+1. Tags - Use values in the [logfmt](https://brandur.org/logfmt) format. For example `error=true db.statement="select * from User"`.
+1. Min Duration - Setting this field should filter traces that has higher duration than the value. Possible values are `1.2s, 100ms, 500us`.
```
```suggestion
1. Min Duration - Filter all traces with a duration higher than the set value. Possible values are `1.2s, 100ms, 500us`.
```

achatterjee-grafana commented on docs/sources/datasources/jaeger.md:
```diff
@@ -39,25 +39,24 @@ This is a configuration for the [trace to logs feature]({{< relref "../explore/t
+1. Operation - Populated after selecting a service. Lists operations related to the selected service. Select [All] option to query all the operation.
+1. Tags - Use values in the [logfmt](https://brandur.org/logfmt) format. For example `error=true db.statement="select * from User"`.
+1. Min Duration - Setting this field should filter traces that has higher duration than the value. Possible values are `1.2s, 100ms, 500us`.
+1. Max Duration - Setting this field should filter traces that duration are not higher than the value. Possible values are `1.2s, 100ms, 500us`.
```
```suggestion
1. Max Duration - Filter all traces with a duration lower than the set value. Possible values are `1.2s, 100ms, 500us`.
```

achatterjee-grafana reviewed (changes requested):
Added copy edit changes.

zoltanbedi committed:
Apply suggestions from code review

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

zoltanbedi committed:
Merge remote-tracking branch 'origin/master' into zoltan/better-jaeger-search

zoltanbedi committed:
Merge branch 'zoltan/better-jaeger-search' of github.com:grafana/grafana into zoltan/better-jaeger-search

zoltanbedi committed:
Update Jaeger doc

achatterjee-grafana reviewed (approved):
LGTM

zoltanbedi committed:
UI updates

zoltanbedi committed:
Merge remote-tracking branch 'origin/master' into zoltan/better-jaeger-search

aocenas commented:
Code looks good but there are some behavior that I think we should try to fix:
1. The limit placeholder is not seen in full so it's not clear what it means, we should probably put longer description in some info tooltip.
![Screenshot from 2021-05-04 15-21-29](https://user-images.githubusercontent.com/1014802/117010017-aaa2fe00-acec-11eb-9714-039adf37d025.png)

2. When querying logs the order is not stable and I get random traces every time. This isn't happening in jaeger ui so I assume there is either some param in the API or they sort the traces by time after they get them.
3. There seems to be some weird behavior with min/max limits but some of that is also present in Jaeger. In Jaeger if I use both min + max it seems like max is ignored. In this PR it seems like max limit is always ignored.

zoltanbedi committed:
Merge remote-tracking branch 'origin/master' into zoltan/better-jaeger-search

zoltanbedi committed:
Address review comments

zoltanbedi commented:
> Code looks good but there are some behavior that I think we should try to fix:
> 
> 1. The limit placeholder is not seen in full so it's not clear what it means, we should probably put longer description in some info tooltip.
> 2. When querying logs the order is not stable and I get random traces every time. This isn't happening in jaeger ui so I assume there is either some param in the API or they sort the traces by time after they get them.
> 3. There seems to be some weird behavior with min/max limits but some of that is also present in Jaeger. In Jaeger if I use both min + max it seems like max is ignored. In this PR it seems like max limit is always ignored.

1. Added the placeholder text to the info tooltip instead.
2. It seems like that they sort by most recent traces so I changed it to do the same. Don't we have a field config related to sorting?
3. I couldn't reproduce this. I see the same behaviour as in the Jaeger UI.

zoltanbedi committed:
Add new screenshots to docs

aocenas commented:
> I couldn't reproduce this. I see the same behaviour as in the Jaeger UI.

hmm yes you are right, but still this seems super weird, I mean this seems like a bug in the Jaeger API but I think this will create some issues for us as people will assume it's a bug in Grafana. @hemdrup @davkal what do you think about this? From perspective of Grafana user this seems broken. Should we allow only one of mix/max at the time? Or some disclaimer?

Another small nitpick:
Would probably change the placeholder to something like `foo, bar, baz` (or some more real tags) to give an example of the syntax (delimiter) as that is more helpful there imho.
![Screenshot from 2021-05-05 15-06-40](https://user-images.githubusercontent.com/1014802/117146124-2ae06680-adb4-11eb-991d-045afe00e411.png)

hemdrup commented:
> Should we allow only one of mix/max at the time? Or some disclaimer?


- @aocenas Great idea on the "tag" example
- @aocenas Not sure what you mean by "only one min/mac" Do you mean multiple allowing filters like? Show me logs between 2-3s + logs between 4-5s?

zoltanbedi committed:
Merge remote-tracking branch 'origin/master' into zoltan/better-jaeger-search

zoltanbedi committed:
Update placeholder text for tags

zoltanbedi committed:
Merge remote-tracking branch 'origin/main' into zoltan/better-jaeger-search

aocenas reviewed (approved):
We can check the min/max issue in later PR
